### ISSUE TITLE
Representing Square 2D arrays

### ISSUE THREAD

**Issue by as56** [kelloggm/checker-framework#140]
Consider a situation in which I would like to convey that a 2D array such as int[][] a is a square matrix, there is no effective way to currently do this using the index checker, this also could be generalized to arrays of larger dimensions like 3D "cubic" arrays and so on.

**Comment by kelloggm** [comment]
This should be possible with `@PolyMinLen`. The array `int @PolyMinLen [] @PolyMinLen [] x`, for instance, is a 2D square array.

**Comment by mernst** [comment]
That type admits non-square arrays, but it may be good enough to prove desired properties about square arrays.

**Comment by kelloggm** [comment]
You're right, my mistake. That probably is sufficient, though. We'd need a `@PolyArrayLen` annotation in the Value Checker if we wanted exactly the behavior I described.

**Comment by mernst** [comment]
You are right:  `@PolyArrayLen` would specify square arrays exactly.
Polymorphic qualifiers are easy to add and the manual encourages them to be included in every type system.  Is there a reason not to add that one?

**Comment by kelloggm** [comment]
I don't know one. @smillst might if one exists.

**Comment by mernst** [comment]
If it's a recommended best practice, and you don't know of a reason not to do it, then I think you should do it (especially if there is a compelling use case, as in this example).  Thanks!

**Comment by smillst** [comment]
Polymorphic annotations only make sense on method declarations or within a method body where the declaration contains a polymorphic annotation.

```
void test(int @PolyArrayLen [] @PolyArrayLen [] array) {...}
```

Unless `-AinvariantArrays` is used,  the above method code be called with `int @ArrayLen(1)[] @ArrayLen(2)[]`.  (`@PolyArrayLen` is replaced with `@ArrayLen({1,2})`, the lub of `@ArrayLen(1)` and `@ArrayLen(2)`.
